include/dlaf/communication/broadcast_panel.h

```diff
@@ -164,50 +146,44 @@ void broadcast(comm::IndexT_MPI rank_root, matrix::Panel<axis, T, D, storage>& p
-  DLAF_ASSERT_MODERATE(panel.rangeEnd() == panelT.rangeEnd(), panel.rangeEnd(), panelT.rangeEnd());
+  DLAF_ASSERT(panel.rangeStart() <= panelT.rangeStart(), panel.rangeStart(), panelT.rangeStart());
+  DLAF_ASSERT(panel.rangeEnd() >= panelT.rangeEnd(), panel.rangeEnd(), panelT.rangeEnd());
 
   // if no panel tiles, just skip it
   if (panel.rangeStart() == panel.rangeEnd())
     return;
 
   // STEP 1
-  constexpr auto comm_dir_step1 = orthogonal(axis);
-  auto& chain_step1 = internal::get_taskchain<comm_dir_step1>(row_task_chain, col_task_chain);
+  auto& chain_step1 = internal::get_taskchain<coord>(row_task_chain, col_task_chain);
 
   broadcast(rank_root, panel, chain_step1);
 
   // STEP 2
-  constexpr auto comm_dir_step2 = orthogonal(axisT);
-  constexpr auto comm_coord_step2 = axisT;
-
-  auto& chain_step2 = internal::get_taskchain<comm_dir_step2>(row_task_chain, col_task_chain);
-
-  const SizeType last_tile = std::max(panelT.rangeStart(), panelT.rangeEnd() - 1);
-  const auto owner = dist.template rankGlobalTile<coordT>(last_tile);
-  const auto range = dist.rankIndex().get(coordT) == owner
-                         ? common::iterate_range2d(*panelT.iteratorLocal().begin(),
-                                                   LocalTileIndex(coordT, panelT.rangeEndLocal() - 1, 1))
-                         : panelT.iteratorLocal();
-
-  for (const auto& indexT : range) {
-    auto [index_diag, owner_diag] = internal::transposedOwner<coordT>(dist, indexT);
-
-    namespace ex = pika::execution::experimental;
-    if (dist.rankIndex().get(coord) == owner_diag) {
-      const auto index_diag_local = dist.template localTileFromGlobalTile<coord>(index_diag);
-      panelT.setTile(indexT, panel.read({coord, index_diag_local}));
-
-      if (dist.commGridSize().get(comm_coord_step2) > 1)
-        ex::start_detached(schedule_bcast_send(chain_step2.exclusive(), panelT.read(indexT)));
-    }
-    else {
-      if (dist.commGridSize().get(comm_coord_step2) > 1)
-        ex::start_detached(schedule_bcast_recv(chain_step2.exclusive(), owner_diag,
-                                               panelT.readwrite(indexT)));
+  auto& chain_step2 = internal::get_taskchain<coordT>(row_task_chain, col_task_chain);
+  const auto& my_rank = dist.rank_index();
+
+  const auto comm_size = dist.commGridSize().template get<coord>();
+
+  for (SizeType k = panelT.rangeStart(); k < panelT.rangeEnd(); ++k) {
+    const auto kk_rank = dist.rank_global_tile({k, k});
+    if (kk_rank.template get<coordT>() == my_rank.template get<coordT>()) {
+      const auto k_local_T = dist.template local_tile_from_global_tile<coordT>(k);
+
+      namespace ex = pika::execution::experimental;
+      if (kk_rank.template get<coord>() == my_rank.template get<coord>()) {
+        const auto k_local = dist.template local_tile_from_global_tile<coord>(k);
+        panelT.setTile({coordT, k_local_T}, panel.read({coord, k_local}));
+
+        if (comm_size > 1)
```

Comment:
Since this condition is the same in both branches, isn't it worth to add an early return if `comm_size == 1`?

Reply:
You have to set the tile (see the instruction before), no matter the size of the communicator.